lib/addressable/template.rb

```diff
@@ -902,15 +902,14 @@ def normalize_value(value)
 
       # Handle unicode normalization
       if value.kind_of?(Array)
-        value.map! { |val| Addressable::IDNA.unicode_normalize_kc(val) }
+        value.map! { |val| normalize_value(val) }
       elsif value.kind_of?(Hash)
         value = value.inject({}) { |acc, (k, v)|
-          acc[Addressable::IDNA.unicode_normalize_kc(k)] =
-            Addressable::IDNA.unicode_normalize_kc(v)
+          acc[normalize_value(k)] = normalize_value(v)
           acc
         }
-      else
-        value = Addressable::IDNA.unicode_normalize_kc(value)
+      elsif value.encoding == Encoding::UTF_8
+        value = value.unicode_normalize(:nfc)
```

Comment:
According to [the RFC](https://www.rfc-editor.org/rfc/rfc6570.txt), this normalization is actually supposed to be NFC form, not NFKC:
> If a value is provided by a user, such as via a data-entry dialog, then the string SHOULD be normalized as Normalization Form C (NFC: Canonical Decomposition, followed by Canonical Composition) prior to being used in expansions by a template processor.

It's easy to do now with ruby so instead of using the global function I'm switching to direct ruby call. Also I skip the normalize if the input string is not Unicode, this is faster and avoids potential exceptions. Finally I used a recursive call for the array and hash keys/values, this way the same behavior (NFC and only normalize if Unicode) is shared between all types.

Reply:
>I skip the normalize if the input string is not Unicode

Is it possible for the input string to be Unicode but not UTF-8? 🤔 (Maybe not a realistic scenario even if it can happen?)

Reply:
Ah, this is in `Addressable::Template`, maybe @therabidbanana want to chime in?

I'm really not familiar with templates but my feeling is that the change is okay.

Reply:
I'm inclined to agree that this should be changed. It looks like there may be a similar issue with RFC 5890?

> A "U-label" is an IDNA-valid string of Unicode characters, in Normalization Form C (NFC) and including at least one non-ASCII character, expressed in a standard Unicode Encoding Form (such as UTF-8).

A subtle change like this probably deserves some extra attention in tests.

Reply:
> I'm inclined to agree that this should be changed

Do you mean changed back to accepting any encoding and treating it as UTF-8? or keeping the change that I made?

> It looks like there may be a similar issue with RFC 5890/5895?

I believe yes if a user passes in a UTC-16 encoded string for domain name, it'll probably raise. If we want to make this more strict in the IDNA part we could accept any input encoding and re-`encode` it as UTF-8 (instead of `force_encoding`). This is much more likely to break existing people code though if there are invalid or inconvertible chars. Probably something we can discuss more in #491

> A subtle change like this probably deserves some extra attention in tests.

Good point I'll add some more tests around this.

Reply:
Ok I've just added two specs covering the new case: non-unicode strings passed which should **not** be normalized or broken. This direct link should work: https://github.com/sporkmonger/addressable/pull/492/files#diff-ebf4cdeb47084eeefe86a7b9e669d9b89adb907a1c5c19c2928233fabb2c4f86

The result is :green_circle: green on this branch.
With `main` branch the result is :red_circle: :
```
  1) Addressable::Template Partial expand with symbols partial expand with unicode values preserves other encodings when input is not unicode
     Failure/Error: unpacked = input.unpack("U*")
     
     ArgumentError:
       malformed UTF-8 character
     # ./lib/addressable/idna/pure.rb:119:in `unpack'
     # ./lib/addressable/idna/pure.rb:119:in `unicode_normalize_kc'
     # ./lib/addressable/template.rb:948:in `normalize_value'
     # ./lib/addressable/template.rb:809:in `block in transform_capture'
     # ./lib/addressable/template.rb:791:in `each'
     # ./lib/addressable/template.rb:791:in `inject'
     # ./lib/addressable/template.rb:791:in `transform_capture'
     # ./lib/addressable/template.rb:750:in `block in transform_partial_capture'
     # ./lib/addressable/template.rb:747:in `each'
     # ./lib/addressable/template.rb:747:in `inject'
     # ./lib/addressable/template.rb:747:in `transform_partial_capture'
     # ./lib/addressable/template.rb:529:in `block in partial_expand'
     # ./lib/addressable/template.rb:528:in `gsub!'
     # ./lib/addressable/template.rb:528:in `partial_expand'
     # ./spec/addressable/template_spec.rb:1017:in `block (4 levels) in <top (required)>'
```

Probably not a big use-case otherwise people would have complained more I suppose? but I thought it was worth "improving" that.

The only case which could be "degraded" by this new code as @dentarg noted is if people pass in actual unicode string wrongly declared as another encoding. In such case they won't be normalized, whereas in the previous code they would have been. I don't think that's big deal as it was kind of a side-effect initially, and restoring this "behavior" will mean breaking compatibility with some other encodings, and for people who are correctly passing their strings encoding on top of it.

@dentarg @sporkmonger does this makes sense?

Reply:
Thanks for being thorough and checking the new test against `main`.

Reply:
So I think this is the key part of the template spec related to expansion methods:
> It is only during the process of template expansion that a string of characters in a URI Template is REQUIRED to be processed as a sequence of Unicode code points.
I read that as probably disallowing the behavior specified by this test unfortunately?

Reply:
> Thanks for being thorough and checking the new test against main.

:+1:

> I read that as probably disallowing the behavior described by this test description?

Good point, as I quickly scanned the spec (which is new to me) I noticed they mentionned some flexibility ("it should be understood that templates occur in practice as a sequence of characters in whatever form or encoding is suitable for the context in which they occur") which is what I expected considering the world of URI so I stopped there. And I missed the part you're showing right after that which says that for **expansion**, unicode is required.

In that case it makes things simpler, so I changed the code to accept any string, force encode it as UTF-8 and normalize it (so similar to `main` behavior). I've updated the specs to match this: https://github.com/sporkmonger/addressable/pull/492/files#diff-ebf4cdeb47084eeefe86a7b9e669d9b89adb907a1c5c19c2928233fabb2c4f86
They now test explicitely that:
1. unicode is normalized even if passed in as the wrong encoding
2. invalid UTF-8 chars are rejected (with an exception) like before.

Another approach could be to try to `encode` the incoming string to UTF-8 instead of just `force_encoding`, meaning a valid ISO string for example, would be converted to a different but valid UTF-8 string. But I think it's better to leave this behavior to the client as it could be too surprising. Also it could break some existing code usage were people already pass valid unicode string in the wrong encoding.

Reply:
@sporkmonger gentle ping :ping_pong: just to see if you agree, I'll start working on the other MR once this one is merged.

Reply:
@dentarg shall we keep waiting for Bob approval on this one? I don't want to be too pushy.

Reply:
@jarthod I can give him a nudge via email

Big disclaimer: I have not read the different RFCs as you two have, but I think your last comment makes sense.

Reply:
Ok thanks, there's no rush for me so better be safe and wait for @sporkmonger approval then, these RFCs are tricky so 2.5 pairs of eyes is always better :)